Modified PeakPickerHiRes to address issue #5687

alhigaylan opened the pull request:
## Description

#5687 Modified code to allow peak core lacking left or right neighbor peaks to be picked as well. Code passed PeakPickerHiRes_test and I applied it to the supplied test.mzML file from issue #5687 (see attached picture below). 


 I want feedback on how I handled spline bisection. From OpenMS 2013 pub "The bisection method is used to find the root of spline derivative. The root gives the m/z value of the peak's apex and thus the position of the centroid representing the peak; the value of the spline at this position yields the corresponding intensity value." 


  In cases where left or right neighbor peaks are missing, I considered creating a 'phantom' peak whose distance from central m/z peak is equal to the distance between existing core neighbor and central peak. I can then feed 'left_peak_mz', 'central_peak_mz', 'right_peak_mz' to Math::spline_bisection method. 

 This throws error because the introduced 'phantom' peak is outside the range of spline obtained from 'CubicSpline2d(peak_raw_data)'. I could try to include the phantom peak to the peak array 'peak_raw_data', but then how to handle its corresponding ion mobility array? 


 For now, I simply assign 'central_peak_mz' value to the missing 'left_peak_mz' or right_peak_mz' and then this gets passed to Math::spline_bisection method. It works because we are still within the spline range. Picture below shows I obtained a correct isotope mass. So is there a need to properly model symmetrical peak shape? 


 
<img width="1247" alt="raw_isotope" src="https://user-images.githubusercontent.com/63276677/225127018-7c9023f1-aa00-4390-bc4d-8d44b87235df.png">


<img width="1248" alt="Picked_isotope" src="https://user-images.githubusercontent.com/63276677/225126380-e2b82ef2-3e29-4ae7-b110-8bda53724db4.png">


 To run PeakPickerHiRes on this data, you have to override centroid check and specify ms_levels. use -force and -algorithm:ms_level 1

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

timosachsenberg commented on src/openms/source/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.cpp:
```diff
@@ -188,11 +188,17 @@ namespace OpenMS
       }
       // MZ spacing sanity checks
       double left_to_central = 0.0, central_to_right = 0.0, min_spacing = 0.0;
+      bool no_left_neighbor = false, no_right_neighbor = false;
```
maybe inverse?
bool no_left_neighbor -> has_left_neighbor
could remove the double negations !no_...

alhigaylan replied:
Sounds good to me! I will revise these when I make an updated commit :)

cbielow commented:
very intersting solution. Thanks!
Just to be sure this works on all platforms and does not regress by future changes, could you please add a test in PeakPickerHiRes_test.cpp which uses the test.mzML to find the peaks in question.
Our current tests obviously do not capture the new behaviour and we should make sure that the algorithm keeps working as expected. Thanks again!

timosachsenberg commented on src/openms/source/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.cpp:
```diff
@@ -347,10 +364,21 @@ namespace OpenMS
+        // We have to add a proper peak to peak_spline and have to decide what ion mobility value to assign to it.
+
+        // for now, simply replace mz value of missing neighbor with central mz peak.
+        left_neighbor_mz = (no_left_neighbor) ? central_peak_mz: left_neighbor_mz;
```
I think this could be fine.
What is repoted if report_FWHM is enabled?

timosachsenberg replied:
https://github.com/OpenMS/OpenMS/blob/16a93d924ac776a730a05ab9f68fd815f5e1599e/src/openms/include/OpenMS/MATH/MISC/SplineBisection.h#L73
with your solution now, it will start in the middle and probably move up to the top where the derivative gets close to zero and yield the correct result.
If you add a phantom peak it might error because it evaluates the derivative exactly at or slightly outside the peak range. 

I think how you do it is fine with our implementation.

alhigaylan replied:
> I think this could be fine.
> What is repoted if report_FWHM is enabled?

Here is a screenshot with the FWHM reported in ppm. 

<img width="1507" alt="Screen Shot 2023-04-06 at 12 59 02 PM" src="https://user-images.githubusercontent.com/63276677/230446056-17337582-e772-46cd-ba36-ceae75d3f9cc.png">

timosachsenberg commented:
@hroest or @alhigaylan can you check the test results / update the tests?

alhigaylan commented:
> There seems to be some test failing because PeakPickerIterative (a tool from hannes) find more peaks: https://cdash.seqan.de/test/49418916
> 
> ```
> 
> Offending lines:			(tab_width = 8, first_column = 1)
> 
> in1:  /home/jenkins/ws/openms/PR/bldtst/79892110/build/src/tests/topp/PeakPickerIterative.mzML.tmp   (line: 159, position/column: 59/81)
> 			<spectrum id="scan=17359" index="0" defaultArrayLength="!
> 			                                                        6" dataProcessingRef="dp_sp_0">
> 
> in2:  /home/jenkins/ws/openms/PR/bldtst/79892110/source/src/tests/topp/PeakPickerIterative_1_output.mzML   (line: 159, position/column: 59/81)
> 			<spectrum id="scan=17359" index="0" defaultArrayLength="!
> 			                                                        3" dataProcessingRef="dp_sp_0">
> 
> 
> Easy Access:
> /home/jenkins/ws/openms/PR/bldtst/79892110/build/src/tests/topp/PeakPickerIterative.mzML.tmp:159:81:
> /home/jenkins/ws/openms/PR/bldtst/79892110/source/src/tests/topp/PeakPickerIterative_1_output.mzML:159:81:
> 
> diff /home/jenkins/ws/openms/PR/bldtst/79892110/build/src/tests/topp/PeakPickerIterative.mzML.tmp /home/jenkins/ws/openms/PR/bldtst/79892110/source/src/tests/topp/PeakPickerIterative_1_output.mzML
> FuzzyDiff took 0.00 s (wall), 0.00 s (CPU), 0.00 s (system), 0.00 s (user); Peak Memory Usage: 26 MB.
> ```

Ok so I had a discussion with Hannes and we could go about this in two ways: Either make this modified PeakPickerHiRes behavior an optional parameter or update output test for PeakPickerIterative to bypass this. 


 I took a look at PeakPickterIterative_1_input.mzML and output. To me the overall LC-MS map looks similar. 

Raw PeakPickerIterative_1_input map
<img width="1511" alt="Screen Shot 2023-04-06 at 12 47 59 PM" src="https://user-images.githubusercontent.com/63276677/230444113-deb480f3-e240-4b99-a1a5-00fca2f40f81.png">


Current PeakPickerIterative_1_output map

<img width="1508" alt="Screen Shot 2023-04-06 at 12 48 11 PM" src="https://user-images.githubusercontent.com/63276677/230444227-61434382-29df-45c9-bee7-33eac7d87f11.png">


 PeakPickerIterative output when applying it to PeakPickerIterative_1_input.mzML in my modified OpenMS environment (modified PeakPickerHiRes)

<img width="1509" alt="Screen Shot 2023-04-06 at 12 48 22 PM" src="https://user-images.githubusercontent.com/63276677/230444372-410b6886-ffc2-4184-b279-39c5a01b1be7.png">



 Since PeakPickerIterative is not commonly used, would updating the test files make sense? If problems occur in the future then we can this modified PeakPickerHiRes optional.

timosachsenberg commented:
I think updating the tests is fine. On first glance it looks like the old code finds more/better peaks (more consistent with the isotopic traces) though.

jpfeuffer commented:
To me, it looks like the new code finds more.

alhigaylan commented:
I made a new commit with the following changes: 
 1. Removed double negation from PeakPickerHiRes. 
 2. Added new test file to test for PeakPickerHiRes new behavior. The new test is titled "PeakPickerHiRes_6_input/output.mzML" 
 3. Swapped PeakPickterIterative_1_output.mzML with the new file generated in my new environment. 

 I also made a follow-up commit because I had a hiccup and deleted some code from CMakeList.txt file in the test directory. Hope I fixed my mess. 


 Also -- I think I was not properly testing PeakPickerHiRes system-wide. It seems like there are other tests that fail. Namely OpenSWATHWorkflow tests and PeakPickerHiRes. One ID_Mapper test failed too. I took a look at one of the failed PeakPickerHiRes tests and it seems like it is the same story as PeakPickerIterative test. The new code find more peaks and fails. 

 I am not sure about OpenSWATH and ID_Mapper test. I will have to take a more look into it. 


<img width="562" alt="Screen Shot 2023-04-16 at 11 12 23 PM" src="https://user-images.githubusercontent.com/63276677/232370036-23ef9a9a-7a5d-4edd-98c7-27ade4488da9.png">

timosachsenberg commented:
I can give it a quick look

timosachsenberg commented:
I can confirm that I also get these failed tests. But to me it looks like it now reports less peaks???
e.g. the new output when applied in test 4
```
gitpod@alhigaylan-openms-6s4abodgb0g:/workspace/OpenMS/build$ grep "defaultArrayLength" /workspace/OpenMS/build/src/tests/topp/PeakPickerHiRes_4.tmp 
                        <chromatogram id="tr1" index="0" defaultArrayLength="2">
                        <chromatogram id="tr3" index="1" defaultArrayLength="3">
                        <chromatogram id="tr2" index="2" defaultArrayLength="1">
                        <chromatogram id="tr4" index="3" defaultArrayLength="2">
                        <chromatogram id="tr5" index="4" defaultArrayLength="2">
```
vs the old data:
```
gitpod@alhigaylan-openms-6s4abodgb0g:/workspace/OpenMS/build$ grep "defaultArrayLength" 
/workspace/OpenMS/src/tests/topp/PeakPickerHiRes_2_output.mzML
                        <chromatogram id="tr1" index="0" defaultArrayLength="2">
                        <chromatogram id="tr3" index="1" defaultArrayLength="4">
                        <chromatogram id="tr2" index="2" defaultArrayLength="1">
                        <chromatogram id="tr4" index="3" defaultArrayLength="2">
                        <chromatogram id="tr5" index="4" defaultArrayLength="3">
```

I think one should inspect test outputs manually and compare it to the old output (e.g., in TOPPView). Check that no strange behavior is introduced (e.g., lost peaks that should be pickerd) and decide if the new output is ok + update the tests.

timosachsenberg commented:
Especially as it seems to return less peaks from chromatograms (as opposed to the expected "more peaks").

poshul commented:
@alhigaylan can we take a look at this when I am in Toronto?
